<font size="8">Homework 2: Mini Project - AirBnb Document Database</font>

Group number: `<X>`

Group members:
1. `<Adriana Costinha> : <20230567>`
2. `<Ana Silva> : <20230577>`
3. `<Alícia Santos> : <20230525>`

The Homework 2 is comprised of two parts:
1. Data modelling
2. Queries to database to answer the questions

<font size="6">1. Data Modelling</font>

<font size="4">Congratulations! You have been successfully hired as the new Data Engineering and Management team in the AirBNB Business Intelligence department. 
The organization is undergoing restructuring as the performance of the previous teams was not believed to be satisfactory.

Before leaving their post, the head of the Data Modelling department left this note of identified issues.

Currently a lot of data related to AirBNB listings is all collected in the same document. While this has some advantages it has also caused issues. The query performance is known to be too slow, so some decisions on hybridization of data models are required. The previous team did not have experience in the application of patterns, which has been identified as a key area to improve performance through the enhancement of the data model. Indexes have also not been previously applied.

The AirBNB application use has been growing rapidly, especially the the number of reviews. Because of that we have been overwriting them regularly, but the business intelligence department will benefit from storing all the Reviews and analysing them over time. 

Finally, there are some mistakes in the data collection : for instance, writing the same data multiple times, and assigning wrong timestamps for when transactions occurred. The new team will need to decide how to address these.   

In your new role, you are expected to think about the database use case in each question, analyse all the information you have, how commonly is this query needed, how heavy will it be on reads and write etc. You should modify the database schema, where you deem appropriate, to optimise for the business use cases. To do so you should use all the tools you have learned in the lab and lectures, especially embedding and linking, indexes, and patterns. This can involve creating new fields, documents, and collections. Indicate which pattern you are applying and your reasoning for doing so in comments, remember to identify any duplication that is required and any risks of staleness. 

Remember to think about the relevancy of all data that will be returned in each case and if this is needed or adding excessive time to return:
1. to streamline the data collection system;
2. to clean up the data and optimise what will be returned for each use case;
3. to use correct patterns to improve speed of typical queries;
4. to ensure that all departments receive correct and relevant information from the database;
5. to provide the data model schema to inform other departments of the changes.

Good practices in our company include:
1. all newly created fields have CAPITALIZED names;
2. all new queries function on the most updated version of the database (if you make multiple changes all your queries still work after the final changes);
3. for some of the queries you will likely want to change the database schema in some way. Each database transformation choice is properly documented using format:
    * `TRANSFORMATION APPLIED: <NAME and DESCRIPTION OF TRANSFORMATION>`
    * `REASONING BEHIND: <WHY YOU DID IT>`
    * `EXPECTED RESULT: <WHAT IS THE (EXPECTED) RESULT OF RUNNING QUERIES GIVEN THIS TRANSFORMATION>`

</font>


**Initial data setup:**

Before starting on the queries below review the data and look to adjust the schema for the typical use case described below.

1) The most typical use case for the database is to show information relating to a property listing to a customer. This is done by a query to the database which returns one of the listing documents. Currently a lot of time is spent when a listing is retrieved from the database to show to a customer. Decide what information should be returned in a typical query and optimise the structure for this use case. For example, we typically only want a sample of reviews but not all reviews (although all reviews ), the customer does not need to know past transaction data, etc. Update the document schema for this typical use case. This may involve the creation of new collections and documents.  

2) Review the data for any errors (such as transactions that don’t fit the listing) or inappropriate duplication and clean up as appropriate. 


<font size="6">2. Uses for the database</font>

<font size="4">Different AirBNB departments require different analytics from our common database.
Here are the list of all specific questions from different departments: 

**Standard Dificulty Questions:**

3)	Once per month we like to reward hosts with recognition. Pick three superhosts who have at least two property listings that can accommodate more than four people?

4)	One of employees is thinking of buying a property to rent out. Which bed type is most common in the listings that have waterfront and a dishwasher in New York?

5)	We are thinking of identifying someone to hire to write review for us professionally. What is the name of the reviewer who left the longest review in New York?

6)	We are thinking of trying to assess the security of different areas based on the security deposit required. What is the biggest and smallest difference between the price and deposit for security per number of visitors staying at the property?

7)	We want to identify areas by whether they are typically used for short breaks, like weekend mini breaks, or whether they are more suitable for long trips. We will use this information to target advertising of different customers. It is not expected this information will change much over time so we won’t look to update it we just would like a current view. What is the average duration of stay (in nights) per type of property per city (you can use the maximum_nights to measure length of stays)? For each property type return the city with the highest and lowest average value.

**Advanced Dificulty Questions (Remember to consider how to optimise the database for these queries):**

8)	We want to have a new webpage for hosts when setting up their account. It will list suggested typical amenities. This data will need to be available every time a host registers a property but is not expected to change very much. The starting point for the list will be all unique amenities currently listed in properties (across all documents). Optimise the database for this use case and show how the data should be queried.

9)	We want to start tracking our reviewers better. We want to create a webpage that shows the top 20 reviewers and the count of the number of reviews of each of these reviewers. This webpage should be kept up to date. It should also have a link to return the number of reviews for a given reviewer ID or Name (show how to query for number of reviews by ID or query quickly).


10)	For each property we store review scores across different metrics (accuracy, check-in, cleanliness etc). We are thinking we may soon add more metrics, although we don’t know what these will be. We want to be able to easily query the average score across all of these metrics, including any new metrics that might be added without changing the query. Adjust the data model so this can be done and show the query for an example property.

11)	We wish to be able to have better access to information about transaction, we wish to develop a search engine that can calculate the average value of transactions in a given period of time quickly for a given property.


12)	We wish to have a summary webpage that displays information about our top destinations. This webpage should display for each of the top 10 cities some basic information about our operations in the area (number of properties by type for example, average price by type) but you can choose the metrics. For each of the top 10 cities it should also provide some basic information about the top 3 properties in each city (price, number of review, whatever you think useful) to show an example of the properties available in the area. We would like to keep this webpage up to date as information changes.

**Database updates:**

After you have completed the above tasks and have optimised your database show how you would complete each of the updates below. You can create whatever fictional data you like for the below. Remeber to cosider how to prevent any data you have previously created from becoming stale.

13)	Add a new property, this property should have a new host, and should be located in one of the top 10 cities. The host selects the top 10 most common amenities to be listed for the property.

14)	Add a new review to this property, the review should be from one of our top 20 reviewers.

15)	Add a new review metric called x_factor to the new document and give a score of 10. Show that the average across all metrics is correctly calculated for this listing, with the query you previously developed.


# Setup

1. run Jupyter 
2. run Docker
3. create MongoDB container in Command Prompt: make sure you don't have 'leftover' container from previous runs

        docker run --name mongodb -d -e MONGO_INITDB_ROOT_USERNAME=AzureDiamond -e MONGO_INITDB_ROOT_PASSWORD=hunter2 -p 27017:27017 mongo

4. run Studio3T

5. create New Connection OR reconnect to existing Connection:

    5.A	In Studio3T create Connection:
            a.	Press Connection -> New Connection
            b.	Insert credentials into URI field: mongodb://AzureDiamond:hunter2@localhost:27017
            c.	Press Test Connection
            d.	Assign a name to the Connection (top empty line) -> Save
            e.	Press Connect
            
    5.B In Studio3T reconnect to Connection:
            a. press Connect (top left corner)
            b. choose MongoDB connection  -> press Connect (MongoDB container must be runnning by this time)
            
6.	In Studio3T import the Database:
            a.	Press Import
            b.	Choose BSON mongodump archive 
            c.	Find your database path
            d.	Select All file formats and choose the database file ‘sampledata.archive’
            e.	Press Run: be patient while the database loads
            f.	Inspect the collections and documents

Done!

In [1]:
# Python Connector

#!pip install pymongo
# or #!conda install -y pymongo

from datetime import datetime
from pprint import pprint
import time
from bson.objectid import ObjectId

from pymongo import MongoClient

user="AzureDiamond"
password="hunter2"
host="localhost"
port="27017"
protocol="mongodb"


client = MongoClient(f"{protocol}://{user}:{password}@{host}:{port}")

# Database check
db = client.sample_airbnb 
print(f"Database info: {db}\n")
db.name 

Database info: Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'sample_airbnb')



'sample_airbnb'

# Loading and Inspecting database


In [2]:
# Collections are inside our Database 'sample_analytics'

collection_list = db.list_collection_names()

print(f"The database contains {len(collection_list)} collections")
print(f"All collections: {collection_list[0:]}")
print(f"Collection {collection_list[0]} contains {db[collection_list[0]].count_documents({})} documents")

The database contains 2 collections
All collections: ['listingsAndReviews', 'listingsAndReviews_new']
Collection listingsAndReviews contains 5555 documents


In [3]:
pprint(db.listingsAndReviews_new.find_one())

{'_id': '10006546',
 'access': 'We are always available to help guests. The house is fully '
           'available to guests. We are always ready to assist guests. when '
           'possible we pick the guests at the airport.  This service transfer '
           'have a cost per person. We will also have service "meal at home" '
           'with a diverse menu and the taste of each. Enjoy the moment!',
 'accommodates': 8,
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
              

In [4]:
pprint(db.listingsAndReviews_new.find_one()['transactions'])

{'bucket_end_date': datetime.datetime(2016, 11, 23, 0, 0),
 'bucket_start_date': datetime.datetime(1979, 10, 4, 0, 0),
 'transaction_count': 21,
 'transactions': [{'date': datetime.datetime(2008, 8, 12, 0, 0),
                   'price': '132.106378168429131392258568666875362396240234375'},
                  {'date': datetime.datetime(1999, 3, 9, 0, 0),
                   'price': '23.70647577039597564407813479192554950714111328125'},
                  {'date': datetime.datetime(2016, 6, 15, 0, 0),
                   'price': '6.88059855597928216042191706947050988674163818359375'},
                  {'date': datetime.datetime(2015, 12, 17, 0, 0),
                   'price': '2.585702399312638899431249228655360639095306396484375'},
                  {'date': datetime.datetime(2015, 2, 3, 0, 0),
                   'price': '2.56992480431307068755586442421190440654754638671875'},
                  {'date': datetime.datetime(1997, 9, 8, 0, 0),
                   'price': '7.251759727567949

Lets confirm if there are incoherences on the timestamps of the transactions. As it can be seen below, there's not.

In [5]:
query = [{
    "$match": {
        '$expr':{
            '$gt':['$transactions.bucket_start_date','$transactions.bucket_end_date']}}
}, {
    '$limit':3
}]

document = db.listingsAndReviews_new.aggregate(query)


result = list(document)

pprint(result)

[]


[PyMongo documentation](https://pymongo.readthedocs.io/en/stable/api/pymongo/collection.html)

### Question 1
The most typical use case for the database is to show information relating to a property listing to a customer. This is done by a query to the database which returns one of the listing documents. Currently a lot of time is spent when a listing is retrieved from the database to show to a customer. Decide what information should be returned in a typical query and optimise the structure for this use case. For example, we typically only want a sample of reviews but not all reviews (although all reviews ), the customer does not need to know past transaction data, etc. Update the document schema for this typical use case. This may involve the creation of new collections and documents.

# **CREATING COLLECTIONS**


**CREATING TRANSACTIONS:**

We have decided to create a collection with transactions per listing. Once each document has referenced the listing_id, we don't need to have this information referenced on the listing collection side.

We have left the `on` in the case that there is already a document of a listing in the transactions, and we want to add a new transaction to its transaction documents on the transaction collection.

In [6]:
query_1 = {"$unwind": "$transactions"}

query_2 = {"$project": {
        "listingId": "$_id",
        "transactions": "$transactions"}}

query_3 = {"$merge": {
        "into": "transactions", 
        "on": "_id", 
        "whenMatched": "replace" }}

pipeline = [query_1, query_2, query_3]
result = db.listingsAndReviews_new.aggregate(pipeline)
pprint(db.transactions.find_one())

{'_id': '10006546',
 'listingId': '10006546',
 'transactions': {'bucket_end_date': datetime.datetime(2016, 11, 23, 0, 0),
                  'bucket_start_date': datetime.datetime(1979, 10, 4, 0, 0),
                  'transaction_count': 21,
                  'transactions': [{'date': datetime.datetime(2008, 8, 12, 0, 0),
                                    'price': '132.106378168429131392258568666875362396240234375'},
                                   {'date': datetime.datetime(1999, 3, 9, 0, 0),
                                    'price': '23.70647577039597564407813479192554950714111328125'},
                                   {'date': datetime.datetime(2016, 6, 15, 0, 0),
                                    'price': '6.88059855597928216042191706947050988674163818359375'},
                                   {'date': datetime.datetime(2015, 12, 17, 0, 0),
                                    'price': '2.585702399312638899431249228655360639095306396484375'},
                          

**CREATING LISTINGS:**

It being a copy of listingsAndReviews.

In [7]:
db.listingsAndReviews_new.aggregate([{'$out':"listings"}])

In [8]:
pprint(db.listings.find_one())

{'_id': '10006546',
 'access': 'We are always available to help guests. The house is fully '
           'available to guests. We are always ready to assist guests. when '
           'possible we pick the guests at the airport.  This service transfer '
           'have a cost per person. We will also have service "meal at home" '
           'with a diverse menu and the taste of each. Enjoy the moment!',
 'accommodates': 8,
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
              

We remove the transactions from the listings collection documents.

In [9]:
db.listings.update_many(
  {},
  { "$unset": { "transactions": "" } }
);

In [10]:
pprint(db.listings.find_one())

{'_id': '10006546',
 'access': 'We are always available to help guests. The house is fully '
           'available to guests. We are always ready to assist guests. when '
           'possible we pick the guests at the airport.  This service transfer '
           'have a cost per person. We will also have service "meal at home" '
           'with a diverse menu and the taste of each. Enjoy the moment!',
 'accommodates': 8,
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
              

We embed the host information. Every field that starts with 'host' will be embeded onto an array called 'host_information'. Since we couldn't use regex for unseting fields, we decided to do a loop instead. This was done in the case of the impossibility of manually select the fields. If so, it would be much preferable, and therefore this code can be found below as well.

Our pipeline took into account the `regex` operator, found in this documentation: https://www.mongodb.com/docs/manual/reference/operator/query/regex/

In [11]:
#Attribute Pattern in listings: host...
cursor = db.listings.find()
field_set = set()
for doc in cursor:
    for key in doc.keys():
        if key.startswith('host'):
            field_set.add(key)

unset_fields = list(field_set)

pipeline = [
    {
        '$set': {
            'host_information': {
                '$arrayToObject': {
                    '$filter': {
                        'input': {
                            '$objectToArray': '$$ROOT'
                        },
                        'as': 'field',
                        'cond': {
                            '$regexMatch': {
                                'input': '$$field.k',
                                'regex': '^host',
                                'options': 'i'  
                            }
                        }
                    }
                }
            }
        }
    },
    {
        '$unset': unset_fields  
    },
    {
        '$merge': {
            'into': 'listings',
            'whenMatched': 'replace'
        }
    }
]

db.listings.aggregate(pipeline)

In [12]:
pprint(db.listings.find_one())

{'_id': '10006546',
 'access': 'We are always available to help guests. The house is fully '
           'available to guests. We are always ready to assist guests. when '
           'possible we pick the guests at the airport.  This service transfer '
           'have a cost per person. We will also have service "meal at home" '
           'with a diverse menu and the taste of each. Enjoy the moment!',
 'accommodates': 8,
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
              

**CREATING REVIEWERS:**

To decide if we should create a reviewers collection, first we checked the maximum reviews per reviewer to make sure the documents wouldn't be too big/heavy. As it can be seen below, the maximum reviews per reviewer is 24. We also confirmed how many reviewers there are in order to have an idea of how many documents this collection would have, and there are 146640 reviewers.

In [13]:
query_1 = {"$unwind": "$reviews"}

query_2 = {"$group": {
        "_id": "$reviews.reviewer_id", 
        "numberOfReviews": {"$sum": 1}}}

query_3 = {"$project": {
    "_id": 0,
    "reviewer_id": "$_id",
    "numberOfReviews": 1}}

query_4 = {"$sort": {"numberOfReviews": -1}}  # Corrected format

query_5 = {"$limit": 1}


pipeline = [query_1, query_2, query_3, query_4, query_5]

result = list(db.listings.aggregate(pipeline))
pprint(result)

[{'numberOfReviews': 24, 'reviewer_id': '20775242'}]


In [14]:
query_1 = {"$unwind": "$reviews"}

query_2 = {"$group": {"_id": "$reviews.reviewer_id"}}

query_3 = {"$count": "Number of Reviewers"}


pipeline = [query_1, query_2, query_3]

result = list(db.listings.aggregate(pipeline))
pprint(result)

[{'Number of Reviewers': 146640}]


We also considered creating this collection but the number of reviews by city. But since each city had too many reviews/documents, we decided against it.

In [15]:
query_1 = {"$unwind": "$reviews"}

query_2 = {"$group": {
                "_id": "$address.market",  
                "numberOfReviews": {"$sum": 1}}}

query_3 = {"$project": {
                "_id": 0,
                "city": "$_id",  
                "numberOfReviews": 1}}

query_4 = {"$sort": {"numberOfReviews": -1}}


pipeline = [query_1, query_2, query_3, query_4]

result = list(db.listings.aggregate(pipeline))
pprint(result)


[{'city': 'New York', 'numberOfReviews': 30934},
 {'city': 'Porto', 'numberOfReviews': 20560},
 {'city': 'Sydney', 'numberOfReviews': 17158},
 {'city': 'Barcelona', 'numberOfReviews': 15742},
 {'city': 'Hong Kong', 'numberOfReviews': 13203},
 {'city': 'Oahu', 'numberOfReviews': 13006},
 {'city': 'Montreal', 'numberOfReviews': 11141},
 {'city': 'Rio De Janeiro', 'numberOfReviews': 8034},
 {'city': 'The Big Island', 'numberOfReviews': 7034},
 {'city': 'Istanbul', 'numberOfReviews': 5675},
 {'city': 'Maui', 'numberOfReviews': 4716},
 {'city': 'Kauai', 'numberOfReviews': 2555},
 {'city': '', 'numberOfReviews': 25},
 {'city': 'Other (International)', 'numberOfReviews': 9}]


In [16]:
query_1 = {"$unwind": "$reviews"} 

query_2 = {"$group": {
                "_id": "$_id",  
                "numberOfReviews": {"$sum": 1}}}
query_3 = {"$project": {
                "_id": 0,
                "property": "$_id",  
                "numberOfReviews": 1}}
query_4 = {"$sort": {"numberOfReviews": -1}}  


pipeline = [query_1, query_2, query_3, query_4]

result = list(db.listings.aggregate(pipeline))
pprint(result)

[{'numberOfReviews': 533, 'property': '4069429'},
 {'numberOfReviews': 469, 'property': '12954762'},
 {'numberOfReviews': 463, 'property': '95560'},
 {'numberOfReviews': 420, 'property': '476983'},
 {'numberOfReviews': 408, 'property': '5283892'},
 {'numberOfReviews': 402, 'property': '2758817'},
 {'numberOfReviews': 399, 'property': '1284759'},
 {'numberOfReviews': 397, 'property': '1482060'},
 {'numberOfReviews': 391, 'property': '127208'},
 {'numberOfReviews': 385, 'property': '11610598'},
 {'numberOfReviews': 380, 'property': '1144090'},
 {'numberOfReviews': 379, 'property': '18173787'},
 {'numberOfReviews': 351, 'property': '941361'},
 {'numberOfReviews': 348, 'property': '16493868'},
 {'numberOfReviews': 320, 'property': '1332929'},
 {'numberOfReviews': 312, 'property': '652063'},
 {'numberOfReviews': 308, 'property': '7536867'},
 {'numberOfReviews': 308, 'property': '1726242'},
 {'numberOfReviews': 304, 'property': '2373963'},
 {'numberOfReviews': 291, 'property': '1206363'},
 {

So to recap, we have decided to create a collection of reviewers due to:

- The number of reviews per reviewer doesn't exceed the number 24 as shown above.
- The fields of the reviews aren't expected to change frequently, since we decided to maintain the review score in the listing documents.

A little overview of the code below: 

We added a new field called `number_of_reviews` that uses the `first` operator to select the value from the first document encountered in each group.

In [17]:
query_1 = {"$unwind": "$reviews"}

query_2 = {"$group": {
                "_id": "$reviews.reviewer_id",
                "reviewer_name": {"$first": "$reviews.reviewer_name"},
                "number_of_reviews": {"$sum": 1},
                "reviews": {"$push": {
                    "comment": "$reviews.comments",
                    "date": "$reviews.date",
                    "listing_id": "$reviews.listing_id"}}}}

query_3 = {"$merge": {
        "into": "reviewers",
        "on": "_id",
        "whenMatched": "replace"}}
    
    
pipeline = [query_1, query_2, query_3]

result = db.listings.aggregate(pipeline)
pprint(db.reviewers.find_one())

{'_id': '10000032',
 'number_of_reviews': 1,
 'reviewer_name': 'Gaurav',
 'reviews': [{'comment': 'Bronwyn met us at the time mutually agreed to show '
                         'us around the house. she was really friendly. We '
                         'were emailed a house manual which had details one '
                         'would require. The neighbourhood was quiet and we '
                         'enjoyed the proximity to city and the convenience to '
                         'coffee shops and bars. Would definitely recommend '
                         'this place to our friends.',
              'date': datetime.datetime(2017, 1, 1, 5, 0),
              'listing_id': '15719888'}]}


**CREATING REVIEWS:**

Update listings with the top 5 reviews

In [18]:
query_1 = {"$unwind": "$reviews"}

query_2 = {"$sort": {"reviews.date": -1}}

query_3 = {"$group": {
                "_id": "$_id",
                "listing_id": {"$first": "$_id"},
                "reviews": {"$push": "$reviews"},
                "number_of_reviews": {"$sum": 1}}}

query_4 = {"$project": {
                "_id": 0,
                "listing_id": 1,
                "reviews": {"$slice": ["$reviews", 5]},
                "number_of_reviews": 1}}

pipeline = [query_1, query_2, query_3, query_4]

recent_reviews_results = db.listings.aggregate(pipeline)

for result in recent_reviews_results:
    listing_id = result['listing_id']
    recent_reviews = result['reviews']
    # Update the original document in the database
    db.listings.update_one({"_id": listing_id}, {"$set": {"reviews": recent_reviews}})
    
pprint(db.listings.find_one())

{'_id': '10006546',
 'access': 'We are always available to help guests. The house is fully '
           'available to guests. We are always ready to assist guests. when '
           'possible we pick the guests at the airport.  This service transfer '
           'have a cost per person. We will also have service "meal at home" '
           'with a diverse menu and the taste of each. Enjoy the moment!',
 'accommodates': 8,
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
              

We also decided to embed all review_scores. Since we couldn't use regex for unseting fields, we decided to do a loop instead. This was done in the case of the impossibility of manually select the fields. If so, it would be much preferable, and therefore this code can be found below as well.

In [19]:
#Creating the collection:
db.listings.aggregate([{'$out':"reviews"}])

pipeline_1 = [{"$project": {
    'review_scores_checkin': 1,
    'review_scores_cleanliness': 1,
    'review_scores_communication': 1,
    'review_scores_location': 1,
    'review_scores_rating': 1,
    'review_scores_value': 1,
    'reviews': 1}}]

# Execute the aggregation pipeline
db.reviews.update_many({}, pipeline_1)

# Print the result
pprint(db.reviews.find_one())

{'_id': '10006546',
 'review_scores_checkin': 10,
 'review_scores_cleanliness': 9,
 'review_scores_communication': 10,
 'review_scores_location': 10,
 'review_scores_rating': 89,
 'review_scores_value': 9,
 'reviews': [{'_id': '403055315',
              'comments': 'The house was extremely well located and Ana was '
                          'able to give us some really great tips on locations '
                          'to have lunch and eat out. The house was perfectly '
                          'clean and the easily able to accommodate 6 people '
                          'despite only having one bathroom. The beds and '
                          'living room were comfortable. \n'
                          '\n'
                          'However, we always felt somewhat on edge in the '
                          'house due to the number of signs posted around the '
                          'kitchen, bedrooms and bathroom about being charged '
                          '15€ for al

### Question 2

Review the data for any errors (such as transactions that don’t fit the listing) or inappropriate duplication and clean up as appropriate.

In [20]:
#Cleaning the listings collection of the duplicated data in it:
query_1 = {"$unset": {"transactions":""}}
query_2 = {"$unset": {"reviews_copy1":""}}
query_3 = {"$unset": {"reviews_copy2":""}}
query_4 = {"$unset": {"reviews_copy3":""}}
query_5 = {"$unset": {"reviews_copy4":""}}

db.listings.update_many({}, query_1)
db.listings.update_many({}, query_2)
db.listings.update_many({}, query_3)
db.listings.update_many({}, query_4)
db.listings.update_many({}, query_5)

pprint(db.listings.find_one())

{'_id': '10006546',
 'access': 'We are always available to help guests. The house is fully '
           'available to guests. We are always ready to assist guests. when '
           'possible we pick the guests at the airport.  This service transfer '
           'have a cost per person. We will also have service "meal at home" '
           'with a diverse menu and the taste of each. Enjoy the moment!',
 'accommodates': 8,
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
              

Applying the **Atributte Pattern** in the reviews collection by putting all the review_scores in one single array:

In [21]:
pipeline = [
    {'$addFields': {
        'review_scores': {
            'checkin': '$review_scores_checkin',
            'cleanliness': '$review_scores_cleanliness',
            'communication': '$review_scores_communication',
            'location': '$review_scores_location',
            'rating': '$review_scores_rating',
            'value': '$review_scores_value'
        }
    }},
    {'$project': {
        'review_scores_checkin': 0,
        'review_scores_cleanliness': 0,
        'review_scores_communication': 0,
        'review_scores_location': 0,
        'review_scores_rating': 0,
        'review_scores_value': 0

    }},
    {'$merge': {
        'into': 'reviews',
        'on': '_id',
        'whenMatched': 'replace' 
    }}
]


result = db.reviews.aggregate(pipeline)

In [22]:
pprint(db.reviews.find_one())

{'_id': '10006546',
 'review_scores': {'checkin': 10,
                   'cleanliness': 9,
                   'communication': 10,
                   'location': 10,
                   'rating': 89,
                   'value': 9},
 'reviews': [{'_id': '403055315',
              'comments': 'The house was extremely well located and Ana was '
                          'able to give us some really great tips on locations '
                          'to have lunch and eat out. The house was perfectly '
                          'clean and the easily able to accommodate 6 people '
                          'despite only having one bathroom. The beds and '
                          'living room were comfortable. \n'
                          '\n'
                          'However, we always felt somewhat on edge in the '
                          'house due to the number of signs posted around the '
                          'kitchen, bedrooms and bathroom about being charged '
            

We noticed that some fields like `maximum_nights` and `minimum_nights` had type string instead of integer, so we updated it.

In [23]:
documents_to_update = {"maximum_nights":{"$type":"string"}}

update_info ={"$set": {
                "maximum_nights": {
                    "$toInt": "$maximum_nights"}
        }
    }

pprint(db.listings.update_many(documents_to_update,[update_info]))

UpdateResult({'n': 5555, 'nModified': 5555, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)


In [24]:
documents_to_update = {"minimum_nights":{"$type":"string"}}

update_info ={"$set": {
                "minimum_nights": {
                    "$toInt": "$minimum_nights"}
        }
    }

pprint(db.listings.update_many(documents_to_update,[update_info]))

UpdateResult({'n': 5555, 'nModified': 5555, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)


### Question 3
Once per month we like to reward hosts with recognition. 
Pick three superhosts who have at least two property listings that can accommodate more than four people?

Needing this information only once per month means little frequent access to the database in order to get this information.
Therefore, we have decided not to create a collection, but rather a pipeline that outputs the requiered result.

We analysed the necessity of creating an index. Without an index, as it can be seen below, there were 5555 documents analysed:  **'totalDocsExamined': 5555**. After creating a **compound index**, there were only 52, so we were able to optimize this query. By adding this index the _write performance_ will suffer, although we found this to not be relevant since this action would not be performed frequently.

In the database there can be more than three superhosts who meet this criteria, therefore we have chose to randomly select three of them.

In [25]:
find = db.listings.find({'host_is_superhost': True, 'accommodates': {"$gt": 4}})
stats = find.explain()
pprint(stats['executionStats'])

{'allPlansExecution': [],
 'executionStages': {'advanced': 0,
                     'direction': 'forward',
                     'docsExamined': 5555,
                     'executionTimeMillisEstimate': 3,
                     'filter': {'$and': [{'host_is_superhost': {'$eq': True}},
                                         {'accommodates': {'$gt': 4}}]},
                     'isEOF': 1,
                     'nReturned': 0,
                     'needTime': 5555,
                     'needYield': 0,
                     'restoreState': 5,
                     'saveState': 5,
                     'stage': 'COLLSCAN',
                     'works': 5556},
 'executionSuccess': True,
 'executionTimeMillis': 21,
 'nReturned': 0,
 'totalDocsExamined': 5555,
 'totalKeysExamined': 0}


In [26]:
db.listings.create_index([('host_is_superhost', 1), ('accommodates', 1)])

'host_is_superhost_1_accommodates_1'

In [27]:
find = db.listings.find({'host_is_superhost': True, 'accommodates': 1})
stats = find.explain()
pprint(stats['executionStats'])

{'allPlansExecution': [],
 'executionStages': {'advanced': 0,
                     'alreadyHasObj': 0,
                     'docsExamined': 0,
                     'executionTimeMillisEstimate': 1,
                     'inputStage': {'advanced': 0,
                                    'direction': 'forward',
                                    'dupsDropped': 0,
                                    'dupsTested': 0,
                                    'executionTimeMillisEstimate': 1,
                                    'indexBounds': {'accommodates': ['[1, 1]'],
                                                    'host_is_superhost': ['[true, '
                                                                          'true]']},
                                    'indexName': 'host_is_superhost_1_accommodates_1',
                                    'indexVersion': 2,
                                    'isEOF': 1,
                                    'isMultiKey': False,
                  

In [28]:
query_1 = {
    '$match': {
        'host_information.host_is_superhost': True,
        'accommodates': {'$gt': 4},
    }
}

query_2 = {
    '$group': {
        '_id': '$host_information.host_id',  
        'numberOfListings': {'$sum': 1},
        'accommodates_max': {'$max': '$accommodates'}
    }
}

query_3 = {
    '$match': {
        'numberOfListings': {'$gt': 2}
    }
}

query_4 = {
    '$project': {
        'host_id': '$_id',
        'numberOfListings': 1,
        'accommodates_max': 1
    }
}

pipeline = [query_1, query_2, query_3, query_4]
result = list(db.listings.aggregate(pipeline))

if len(result) > 3:
    selected_hosts = random.sample(result, 3)
else:
    selected_hosts = result

pprint(selected_hosts)

[{'_id': '10496350',
  'accommodates_max': 14,
  'host_id': '10496350',
  'numberOfListings': 3},
 {'_id': '42618840',
  'accommodates_max': 6,
  'host_id': '42618840',
  'numberOfListings': 3},
 {'_id': '36133410',
  'accommodates_max': 8,
  'host_id': '36133410',
  'numberOfListings': 3}]


### Question 4
One of employees is thinking of buying a property to rent out. Which bed type is most common in the listings that have waterfront and a dishwasher in New York?

To start answering this question, we analysed the necessity of creating an index. Without an index, as it can be seen below, there were 5555 documents analysed:  **'totalDocsExamined': 5555**. After creating it, there were only 75, so we were able to optimize this query and collection. By adding this **compound index** the _write performance_ will suffer, although we found this to not be relevant since this action would not be performed frequently.

Then we simply applied this index in order to get the answer.

In [29]:
pprint(db.listings.index_information())

{'_id_': {'key': [('_id', 1)], 'v': 2},
 'host_is_superhost_1_accommodates_1': {'key': [('host_is_superhost', 1),
                                                ('accommodates', 1)],
                                        'v': 2}}


In [30]:
find = db.listings.find({'amenities': { '$in': ['Waterfront', 'Dishwasher']}, 'address.market': 'New York'})
stats = find.explain()
pprint(stats['executionStats'])

{'allPlansExecution': [],
 'executionStages': {'advanced': 75,
                     'direction': 'forward',
                     'docsExamined': 5555,
                     'executionTimeMillisEstimate': 26,
                     'filter': {'$and': [{'address.market': {'$eq': 'New '
                                                                    'York'}},
                                         {'amenities': {'$in': ['Dishwasher',
                                                                'Waterfront']}}]},
                     'isEOF': 1,
                     'nReturned': 75,
                     'needTime': 5480,
                     'needYield': 0,
                     'restoreState': 5,
                     'saveState': 5,
                     'stage': 'COLLSCAN',
                     'works': 5556},
 'executionSuccess': True,
 'executionTimeMillis': 47,
 'nReturned': 75,
 'totalDocsExamined': 5555,
 'totalKeysExamined': 0}


In [31]:
db.listings.create_index([('amenities', 1), ('address.market', 1)])

'amenities_1_address.market_1'

In [32]:
find = db.listings.find({'amenities': { '$in': ['Waterfront', 'Dishwasher']}, 'address.market': 'New York'})
stats = find.explain()
pprint(stats['executionStats'])

{'allPlansExecution': [],
 'executionStages': {'advanced': 75,
                     'alreadyHasObj': 0,
                     'docsExamined': 75,
                     'executionTimeMillisEstimate': 1,
                     'inputStage': {'advanced': 75,
                                    'direction': 'forward',
                                    'dupsDropped': 1,
                                    'dupsTested': 76,
                                    'executionTimeMillisEstimate': 1,
                                    'indexBounds': {'address.market': ['["New '
                                                                       'York", '
                                                                       '"New '
                                                                       'York"]'],
                                                    'amenities': ['["Dishwasher", '
                                                                  '"Dishwasher"]',
                     

In [33]:
query_1 = {
          '$match': {
                'amenities': { '$all': ['Waterfront', 'Dishwasher'] }, 'address.market': 'New York'
  }
}

query_2 = {"$group": {
        "_id": "$bed_type",
        "number_of_bed_types": {"$sum": 1}
    }
}


r = db.listings.aggregate([query_1, query_2])

result = list(r)

pprint(result)

[{'_id': 'Real Bed', 'number_of_bed_types': 1}]


### Question 5

We are thinking of identifying someone to hire to write review for us professionally. What is the name of the reviewer who left the longest review in New York?

When writing the code we came across an error while trying to calculate the length of the comment: `OperationFailure: PlanExecutor error during aggregation :: caused by :: $strLenCP requires a string argument, found: missing`.
This could mean one of two things:
1. Either the comment field was missing
2. Or the comment field was simply null

In order to resolve this issue we investigated further, and found indeed reviews that had no comments attached. 

In [34]:
query_1 = {'$unwind': '$reviews'}

query_2 ={'$match': {
        '$or': [{'reviews.comment': {'$eq': None}}, 
                {'reviews.comment': {'$exists': False}},
               {'address.market': 'New York'}]}
         }

query_3 = {'$project': {
                '_id': 1,
                'reviewer_name': 1,
                'reviews': 1}
          }


pipeline = [query_1, query_2, query_3]

result = list(db.reviewers.aggregate(pipeline))
for doc in result:
    pprint(doc)

{'_id': '120904349',
 'reviewer_name': 'Joshua',
 'reviews': {'date': datetime.datetime(2017, 4, 5, 4, 0),
             'listing_id': '3961051'}}
{'_id': '129030966',
 'reviewer_name': 'Edward',
 'reviews': {'date': datetime.datetime(2017, 5, 9, 4, 0),
             'listing_id': '16389574'}}
{'_id': '136463544',
 'reviewer_name': 'Dennis',
 'reviews': {'date': datetime.datetime(2017, 12, 1, 5, 0),
             'listing_id': '260703'}}
{'_id': '227693876',
 'reviewer_name': 'Joseph',
 'reviews': {'date': datetime.datetime(2018, 11, 30, 5, 0),
             'listing_id': '18914730'}}
{'_id': '52185090',
 'reviewer_name': 'Marc',
 'reviews': {'date': datetime.datetime(2017, 3, 5, 5, 0),
             'listing_id': '15386249'}}
{'_id': '57472804',
 'reviewer_name': 'Janice',
 'reviews': {'date': datetime.datetime(2017, 7, 23, 4, 0),
             'listing_id': '15285714'}}
{'_id': '7323479',
 'reviewer_name': 'Zachary',
 'reviews': {'date': datetime.datetime(2018, 10, 21, 4, 0),
             

Therefore we added a condition: 
*  _if the comment field was non existing or null_ - we would replace this field with a missing string in order for its length to be zero. 

In [35]:
find = db.listings.find({'address.market': 'New York'})
stats = find.explain()
pprint(stats['executionStats'])

{'allPlansExecution': [],
 'executionStages': {'advanced': 607,
                     'direction': 'forward',
                     'docsExamined': 5555,
                     'executionTimeMillisEstimate': 4,
                     'filter': {'address.market': {'$eq': 'New York'}},
                     'isEOF': 1,
                     'nReturned': 607,
                     'needTime': 4948,
                     'needYield': 0,
                     'restoreState': 5,
                     'saveState': 5,
                     'stage': 'COLLSCAN',
                     'works': 5556},
 'executionSuccess': True,
 'executionTimeMillis': 27,
 'nReturned': 607,
 'totalDocsExamined': 5555,
 'totalKeysExamined': 0}


In [36]:
db.listings.create_index([('address.market', 1)])

'address.market_1'

In [37]:
find = db.listings.find({'address.market': 'New York'})
stats = find.explain()
pprint(stats['executionStats'])

{'allPlansExecution': [],
 'executionStages': {'advanced': 607,
                     'alreadyHasObj': 0,
                     'docsExamined': 607,
                     'executionTimeMillisEstimate': 0,
                     'inputStage': {'advanced': 607,
                                    'direction': 'forward',
                                    'dupsDropped': 0,
                                    'dupsTested': 0,
                                    'executionTimeMillisEstimate': 0,
                                    'indexBounds': {'address.market': ['["New '
                                                                       'York", '
                                                                       '"New '
                                                                       'York"]']},
                                    'indexName': 'address.market_1',
                                    'indexVersion': 2,
                                    'isEOF': 1,
             

In [38]:
pipeline = [
    {'$lookup': {
        'from': 'listings',
        'localField': '_id',
        'foreignField': '_id',
        'as': 'listings'
    }},
    {'$unwind': '$listings'},
    {'$match': {'listings.address.market': 'New York'}},
    {'$unwind': '$reviews'},
    {'$match': {'reviews.comments': {'$exists': True, '$ne': ''}}},
    {'$addFields': {
        'review_length': {'$strLenCP': '$reviews.comments'}
    }},
    {'$sort': {'review_length': -1}},
    {'$project': {
        '_id': 0,
        'review_name': '$reviews.reviewer_name',
        'review_length': 1
    }},
    {'$limit': 1}
]

result = list(db.reviews.aggregate(pipeline))
pprint(result)

[{'review_length': 4665, 'review_name': 'Angela'}]


### Question 6
We are thinking of trying to assess the security of different areas based on the security deposit required. What is the biggest and smallest difference between the price and deposit for security per number of visitors staying at the property?

In order of assessing the security of different areas based on the difference between price and security deposity, we consider the city of each listing moving forward. 

Firstly, we start by grouping the city and the number of visitors, and calculating the highest and lowest difference between price and security deposity. Getting an output of the differences per city and  'accommodates'.

We realise there were differences that had a 'None' value, so with a `match` operator we exclude those cases. 

Lastly we group the output by city and organize it with a `push` operator.

In [39]:
query_1 = {'$match': {'address.market': {'$ne': ''}}}

query_2 = {'$group':{
                '_id': {'city': '$address.market', 'accommodates': '$accommodates'}, #need to be both ids, otherwise it wont group the number of people
                'max_dif': {'$max': {'$abs': {'$subtract': ['$price', '$security_deposit']}}},
                'min_dif': {'$min': {'$abs': {'$subtract': ['$price', '$security_deposit']}}}
    }}

query_3 = {'$match': {
                'max_dif': {'$ne': None},
                'min_dif': {'$ne': None}
    }}

query_4 = {'$group': {
                '_id': '$_id.city',
                'differences': {
                    '$push': {
                        'accommodates': '$_id.accommodates',
                        'max_dif': '$max_dif',
                        'min_dif': '$min_dif'}}
    }}

pipeline=[query_1, query_2, query_3, query_4]

result = db.listings.aggregate(pipeline)
pprint(list(result))

[{'_id': 'Porto',
  'differences': [{'accommodates': 8,
                   'max_dif': Decimal128('300.00'),
                   'min_dif': Decimal128('15.00')},
                  {'accommodates': 1,
                   'max_dif': Decimal128('91.00'),
                   'min_dif': Decimal128('13.00')},
                  {'accommodates': 6,
                   'max_dif': Decimal128('620.00'),
                   'min_dif': Decimal128('10.00')},
                  {'accommodates': 12,
                   'max_dif': Decimal128('209.00'),
                   'min_dif': Decimal128('60.00')},
                  {'accommodates': 10,
                   'max_dif': Decimal128('320.00'),
                   'min_dif': Decimal128('60.00')},
                  {'accommodates': 3,
                   'max_dif': Decimal128('462.00'),
                   'min_dif': Decimal128('25.00')},
                  {'accommodates': 4,
                   'max_dif': Decimal128('925.00'),
                   'min_dif': Decimal12

                  {'accommodates': 3,
                   'max_dif': Decimal128('2849.00'),
                   'min_dif': Decimal128('11.00')},
                  {'accommodates': 6,
                   'max_dif': Decimal128('1601.00'),
                   'min_dif': Decimal128('0.00')},
                  {'accommodates': 7,
                   'max_dif': Decimal128('2501.00'),
                   'min_dif': Decimal128('15.00')},
                  {'accommodates': 1,
                   'max_dif': Decimal128('788.00'),
                   'min_dif': Decimal128('43.00')},
                  {'accommodates': 15,
                   'max_dif': Decimal128('40.00'),
                   'min_dif': Decimal128('40.00')},
                  {'accommodates': 14,
                   'max_dif': Decimal128('811.00'),
                   'min_dif': Decimal128('372.00')}]}]


### Question 7
We want to identify areas by whether they are typically used for short breaks, like weekend mini breaks, or whether they are more suitable for long trips. We will use this information to target advertising of different customers. It is not expected this information will change much over time so we won’t look to update it we just would like a current view. What is the average duration of stay (in nights) per type of property per city (you can use the maximum_nights to measure length of stays)? For each property type return the city with the highest and lowest average value.

**Average stay per city**

- Started by grouping the documents per property type and after per city (address.market).
- With the `group` function we created a new field called `average_stay` that returns the average maximum nights per city per property type.
- With `project` we return the dictionary with this new organization and after we user `sort` so we can properly see the distribution per property type.

In [40]:
#Average stay per city
query_1 = {"$group":{
        "_id":{"key1":"$property_type", "key2":"$address.market"},
        "average_stay":{"$avg":"$maximum_nights"}
    }}

query_2 = {"$project": {
        "_id": 0,
        "property_type":"$_id.key1",
        "city":"$_id.key2",
        "average_stay":1,
    }}

query_3 = {"$sort": {
        "property_type":1,
        "average_stay":-1
    }}

pipeline = [query_1, query_2, query_3];

result = list(db.listings.aggregate(pipeline))
pprint(result)

[{'average_stay': 1125.0, 'city': 'Montreal', 'property_type': 'Aparthotel'},
 {'average_stay': 1125.0, 'city': 'New York', 'property_type': 'Aparthotel'},
 {'average_stay': 804.6666666666666,
  'city': 'Istanbul',
  'property_type': 'Aparthotel'},
 {'average_stay': 780.0, 'city': 'Barcelona', 'property_type': 'Aparthotel'},
 {'average_stay': 212.5,
  'city': 'Rio De Janeiro',
  'property_type': 'Aparthotel'},
 {'average_stay': 60.0, 'city': 'Oahu', 'property_type': 'Aparthotel'},
 {'average_stay': 5200437.653753026,
  'city': 'Istanbul',
  'property_type': 'Apartment'},
 {'average_stay': 4837511.168918919,
  'city': 'Hong Kong',
  'property_type': 'Apartment'},
 {'average_stay': 4392173.707566462,
  'city': 'New York',
  'property_type': 'Apartment'},
 {'average_stay': 2650012.9549356224,
  'city': 'Montreal',
  'property_type': 'Apartment'},
 {'average_stay': 1359.252808988764,
  'city': 'Barcelona',
  'property_type': 'Apartment'},
 {'average_stay': 832.9, 'city': 'Kauai', 'property

 {'average_stay': 1125.0,
  'city': 'Other (International)',
  'property_type': 'House'},
 {'average_stay': 1088.909090909091, 'city': 'Maui', 'property_type': 'House'},
 {'average_stay': 863.125, 'city': 'Barcelona', 'property_type': 'House'},
 {'average_stay': 837.0, 'city': 'Istanbul', 'property_type': 'House'},
 {'average_stay': 756.0769230769231,
  'city': 'Hong Kong',
  'property_type': 'House'},
 {'average_stay': 721.068493150685, 'city': 'Porto', 'property_type': 'House'},
 {'average_stay': 710.702380952381, 'city': 'Sydney', 'property_type': 'House'},
 {'average_stay': 675.9324324324324, 'city': 'Oahu', 'property_type': 'House'},
 {'average_stay': 658.9107142857143,
  'city': 'Rio De Janeiro',
  'property_type': 'House'},
 {'average_stay': 624.6590909090909,
  'city': 'Montreal',
  'property_type': 'House'},
 {'average_stay': 621.5, 'city': 'The Big Island', 'property_type': 'House'},
 {'average_stay': 573.25, 'city': 'Kauai', 'property_type': 'House'},
 {'average_stay': 572.3

- We notice that in some cities the average stay is **greater that 365** which we conclude those cases  may be refering to an eventual rent. We consider that we are analysing per year, so we recode it, but this time only with listings that have ``maximum_nights`` less then or equal to 365.

In [41]:
query_0 = {"$match": {
        "maximum_nights": {"$lte": 365}
    }}

query_1 = {"$group":{
        "_id":{"key1":"$property_type", "key2":"$address.market"},
        "average_stay":{"$avg":"$maximum_nights"}
    }}

query_2 = {"$project": {
        "_id": 0,
        "property_type":"$_id.key1",
        "city":"$_id.key2",
        "average_stay":1,
    }}

query_3 = {"$sort": {
        "property_type":1,
        "average_stay":-1
    }}

pipeline = [query_0, query_1, query_2, query_3];

result = list(db.listings.aggregate(pipeline))
pprint(result)

[{'average_stay': 212.5,
  'city': 'Rio De Janeiro',
  'property_type': 'Aparthotel'},
 {'average_stay': 164.0, 'city': 'Istanbul', 'property_type': 'Aparthotel'},
 {'average_stay': 90.0, 'city': 'Barcelona', 'property_type': 'Aparthotel'},
 {'average_stay': 60.0, 'city': 'Oahu', 'property_type': 'Aparthotel'},
 {'average_stay': 151.33333333333334,
  'city': 'Kauai',
  'property_type': 'Apartment'},
 {'average_stay': 122.12, 'city': 'Oahu', 'property_type': 'Apartment'},
 {'average_stay': 117.63546798029557,
  'city': 'Barcelona',
  'property_type': 'Apartment'},
 {'average_stay': 112.97744360902256,
  'city': 'Hong Kong',
  'property_type': 'Apartment'},
 {'average_stay': 88.75,
  'city': 'The Big Island',
  'property_type': 'Apartment'},
 {'average_stay': 83.29113924050633,
  'city': 'Istanbul',
  'property_type': 'Apartment'},
 {'average_stay': 80.21428571428571,
  'city': 'Maui',
  'property_type': 'Apartment'},
 {'average_stay': 79.00408163265305,
  'city': 'New York',
  'property

**Highest and lowest average stay per property type**

- Using the **previous code** we look for the `first` and `last` observation per property type, considering the first the highest and the last the lowest since we sorted the values by property_type ascending and average_stay descending.

In [42]:
query_4 = {"$group": {
    "_id": "$property_type",
    "max_city": {"$first": "$city"},
    "max_avg_stay": {"$first": "$average_stay"},
    "min_city": {"$last": "$city"},
    "min_avg_stay": {"$last": "$average_stay"}
}}

query_5 = {"$project": {
    "_id": 0,
    "property_type": "$_id",
    "max_city": 1,
    "max_avg_stay": 1,
    "min_city": 1,
    "min_avg_stay": 1
}}

query_6 = {"$sort": {
        "property_type":1,
    }}

pipeline = [query_1, query_2, query_3, query_4, query_5, query_6]
result = list(db.listings.aggregate(pipeline))
pprint(result)

[{'max_avg_stay': 1125.0,
  'max_city': 'Montreal',
  'min_avg_stay': 60.0,
  'min_city': 'Oahu',
  'property_type': 'Aparthotel'},
 {'max_avg_stay': 5200437.653753026,
  'max_city': 'Istanbul',
  'min_avg_stay': 371.3636363636364,
  'min_city': 'The Big Island',
  'property_type': 'Apartment'},
 {'max_avg_stay': 1125.0,
  'max_city': 'The Big Island',
  'min_avg_stay': 1125.0,
  'min_city': 'The Big Island',
  'property_type': 'Barn'},
 {'max_avg_stay': 1125.0,
  'max_city': 'Other (International)',
  'min_avg_stay': 30.0,
  'min_city': 'Kauai',
  'property_type': 'Bed and breakfast'},
 {'max_avg_stay': 1125.0,
  'max_city': 'Barcelona',
  'min_avg_stay': 1125.0,
  'min_city': 'Porto',
  'property_type': 'Boat'},
 {'max_avg_stay': 1125.0,
  'max_city': 'Kauai',
  'min_avg_stay': 577.5,
  'min_city': 'Oahu',
  'property_type': 'Boutique hotel'},
 {'max_avg_stay': 1125.0,
  'max_city': 'Montreal',
  'min_avg_stay': 485.0,
  'min_city': 'Oahu',
  'property_type': 'Bungalow'},
 {'max_avg_

### Question 8

We want to have a new webpage for hosts when setting up their account. It will list suggested typical amenities. This data will need to be available every time a host registers a property but is not expected to change very much. The starting point for the list will be all unique amenities currently listed in properties (across all documents). Optimise the database for this use case and show how the data should be queried.

Since this is information that is not expected to change very much, we have decided to create a collection made of unique amenities as documents, with a field which indicates how many properties have this amenitiy. Therefore, it will be easier to retrieve this information. 

In [43]:
query_1 = {"$unwind": "$amenities"}

query_2 = {"$group": {
        "_id": "$amenities",  
        "count": {"$sum": 1}  
}}

query_3 = {"$project": {
        "_id": 0,
        "amenityName": "$_id",
        "count": 1  
}}
query_4 = {"$sort": {"count": -1}}

query_5 = {"$merge": {
        "into": "amenities",  
        "whenMatched": "replace",  
        "whenNotMatched": "insert"  
}}
pipeline = [query_1, query_2, query_3, query_4, query_5]


db.listings.aggregate(pipeline)

result = db.amenities.find()
pprint(list(result))


[{'_id': ObjectId('662598a65718344933ddc4de'),
  'amenityName': 'Wifi',
  'count': 5303},
 {'_id': ObjectId('662598a65718344933ddc4df'),
  'amenityName': 'Essentials',
  'count': 5048},
 {'_id': ObjectId('662598a65718344933ddc4e0'),
  'amenityName': 'Kitchen',
  'count': 4951},
 {'_id': ObjectId('662598a65718344933ddc4e1'),
  'amenityName': 'TV',
  'count': 4295},
 {'_id': ObjectId('662598a65718344933ddc4e2'),
  'amenityName': 'Hangers',
  'count': 4226},
 {'_id': ObjectId('662598a65718344933ddc4e3'),
  'amenityName': 'Hair dryer',
  'count': 3900},
 {'_id': ObjectId('662598a65718344933ddc4e4'),
  'amenityName': 'Washer',
  'count': 3877},
 {'_id': ObjectId('662598a65718344933ddc4e5'),
  'amenityName': 'Shampoo',
  'count': 3709},
 {'_id': ObjectId('662598a65718344933ddc4e6'),
  'amenityName': 'Iron',
  'count': 3692},
 {'_id': ObjectId('662598a65718344933ddc4e7'),
  'amenityName': 'Laptop friendly workspace',
  'count': 3442},
 {'_id': ObjectId('662598a65718344933ddc4e8'),
  'amenityN

### Question 9

We want to start tracking our reviewers better. We want to create a webpage that shows the top 20 reviewers and the count of the number of reviews of each of these reviewers. This webpage should be kept up to date. It should also have a link to return the number of reviews for a given reviewer ID or Name (show how to query for number of reviews by ID or query quickly).

Previously we created a collection called `reviwers` where group all the comments that a reviewer has done.

In this collection we also created a new field `number_of_reviews` that returns the number of comments made by user.

Therefore we just use the `project` function to return that field, `sort` the values by number_of_reviews descending and `limit` to show only the fist 20 that will be consider the top 20 reviewers.

In [44]:
query_1 = {"$project": {
             "_id":1,
              "reviewer_name":1,
              "number_of_reviews":1
           }}

query_2 = {"$sort":{
              "number_of_reviews":-1,
              "_id":1
          }}

query_3 = {"$limit":20}

pipeline = [query_1, query_2, query_3]

result = list(db.reviewers.aggregate(pipeline))

pprint(result)

[{'_id': '20775242', 'number_of_reviews': 24, 'reviewer_name': 'Filipe'},
 {'_id': '67084875', 'number_of_reviews': 13, 'reviewer_name': 'Nick'},
 {'_id': '2961855', 'number_of_reviews': 10, 'reviewer_name': 'Uge'},
 {'_id': '162027327', 'number_of_reviews': 9, 'reviewer_name': 'Thien'},
 {'_id': '20991911', 'number_of_reviews': 9, 'reviewer_name': 'Lisa'},
 {'_id': '12679057', 'number_of_reviews': 8, 'reviewer_name': 'Jodi'},
 {'_id': '1705870', 'number_of_reviews': 8, 'reviewer_name': 'David'},
 {'_id': '55241576', 'number_of_reviews': 8, 'reviewer_name': 'Courtney'},
 {'_id': '60816198', 'number_of_reviews': 8, 'reviewer_name': 'Todd'},
 {'_id': '69140895', 'number_of_reviews': 8, 'reviewer_name': 'Lisa'},
 {'_id': '47303133', 'number_of_reviews': 7, 'reviewer_name': 'Lance'},
 {'_id': '78093968', 'number_of_reviews': 7, 'reviewer_name': 'David'},
 {'_id': '128210181', 'number_of_reviews': 6, 'reviewer_name': 'Branden'},
 {'_id': '24667379', 'number_of_reviews': 6, 'reviewer_name': 

Here is a demonstration of how to query for number of reviews by ID. 

In [45]:
# show number of reviews per id
id = "55241576"

query_1 = {"$project": {
             "_id":1,
              "reviewer_name":1,
              "number_of_reviews":1
           }}

query_2 = {"$sort":{
              "number_of_reviews":-1,
              "_id":1
          }}

query_3 = {"$match":{"_id": id}}


pipeline = [query_1, query_2, query_3]

result = list(db.reviewers.aggregate(pipeline))
pprint(result)

[{'_id': '55241576', 'number_of_reviews': 8, 'reviewer_name': 'Courtney'}]


And a demonstration of how to query for number of reviews by name.

In [46]:
name = "Lisa"

query_1 = {"$project": {
             "_id":1,
              "reviewer_name":1,
              "number_of_reviews":1
           }}

query_2 = {"$sort":{
              "number_of_reviews":-1,
              "_id":1
          }}

query_3 = {"$match":{"reviewer_name": name}}


pipeline = [query_1, query_2, query_3]

result = list(db.reviewers.aggregate(pipeline))
pprint(result)

[{'_id': '20991911', 'number_of_reviews': 9, 'reviewer_name': 'Lisa'},
 {'_id': '69140895', 'number_of_reviews': 8, 'reviewer_name': 'Lisa'},
 {'_id': '1864051', 'number_of_reviews': 2, 'reviewer_name': 'Lisa'},
 {'_id': '233336', 'number_of_reviews': 2, 'reviewer_name': 'Lisa'},
 {'_id': '32826397', 'number_of_reviews': 2, 'reviewer_name': 'Lisa'},
 {'_id': '40437950', 'number_of_reviews': 2, 'reviewer_name': 'Lisa'},
 {'_id': '48811116', 'number_of_reviews': 2, 'reviewer_name': 'Lisa'},
 {'_id': '53968400', 'number_of_reviews': 2, 'reviewer_name': 'Lisa'},
 {'_id': '71136744', 'number_of_reviews': 2, 'reviewer_name': 'Lisa'},
 {'_id': '7494432', 'number_of_reviews': 2, 'reviewer_name': 'Lisa'},
 {'_id': '10053934', 'number_of_reviews': 1, 'reviewer_name': 'Lisa'},
 {'_id': '10263936', 'number_of_reviews': 1, 'reviewer_name': 'Lisa'},
 {'_id': '103350584', 'number_of_reviews': 1, 'reviewer_name': 'Lisa'},
 {'_id': '10385604', 'number_of_reviews': 1, 'reviewer_name': 'Lisa'},
 {'_id': 

### Question 10

For each property we store review scores across different metrics (accuracy, check-in, cleanliness etc). We are thinking we may soon add more metrics, although we don’t know what these will be. We want to be able to easily query the average score across all of these metrics, including any new metrics that might be added without changing the query. Adjust the data model so this can be done and show the query for an example property.

For this question we will use another collection that we created previously.

The collection `reviews` that groups all the comments made on a property and it also contains the reviews_scores metrics associated to a property.

In [47]:
pprint(db.reviews.find_one())

{'_id': '10006546',
 'review_scores': {'checkin': 10,
                   'cleanliness': 9,
                   'communication': 10,
                   'location': 10,
                   'rating': 89,
                   'value': 9},
 'reviews': [{'_id': '403055315',
              'comments': 'The house was extremely well located and Ana was '
                          'able to give us some really great tips on locations '
                          'to have lunch and eat out. The house was perfectly '
                          'clean and the easily able to accommodate 6 people '
                          'despite only having one bathroom. The beds and '
                          'living room were comfortable. \n'
                          '\n'
                          'However, we always felt somewhat on edge in the '
                          'house due to the number of signs posted around the '
                          'kitchen, bedrooms and bathroom about being charged '
            

In [48]:
document = db.reviews.find_one()
for field, value in document.items():
    print(f"Field: {field}, Type: {type(value).__name__}")

Field: _id, Type: str
Field: reviews, Type: list
Field: review_scores, Type: dict


Calculating the average score for a property_id:

In [49]:
property_id = '10006546'

pipeline_10 = [
    {'$match': {'_id': property_id}},  # Filters documents to the specified property
    {'$project': {
        '_id': 1,
        'score_array': {'$objectToArray': '$review_scores'}
    }},
    {'$project': {
        '_id': 1,
        'average_score': {'$avg': '$score_array.v'}
    }}
]

result = list(db.reviews.aggregate(pipeline_10))
pprint(result)

[{'_id': '10006546', 'average_score': 22.833333333333332}]


### Question 11

We wish to be able to have better access to information about transaction, we wish to develop a search engine that can calculate the average value of transactions in a given period of time quickly for a given property.

We started by evaluating the possibility of creating an index capable of searching the date inside the arrays of each property_id. As it can be seen below, it wasn't really necessary, therefore we dropped it.

In order to calculate the average value of transactions in a given period of time for a given property, we first search for the `id_property` with a `match` operator and then after getting the period of time asked with a greater than, `gt`, and a less than , `lt`, operators, we simply calculate the average of the price stated in each transaction for the period of time asked. Finally, we used a `project` operator in order of only showing the necessary informartion.

In [50]:
id_property = '10006546'
start_date = datetime(1997, 1, 1, 0, 0)
end_date = datetime(2001, 12, 31, 0, 0)

pprint(db.transactions.find({'transactions.transactions.date':{'$gt': start_date,'$lt': end_date}}).explain()["executionStats"])

{'allPlansExecution': [],
 'executionStages': {'advanced': 5120,
                     'direction': 'forward',
                     'docsExamined': 5555,
                     'executionTimeMillisEstimate': 20,
                     'filter': {'$and': [{'transactions.transactions.date': {'$lt': datetime.datetime(2001, 12, 31, 0, 0)}},
                                         {'transactions.transactions.date': {'$gt': datetime.datetime(1997, 1, 1, 0, 0)}}]},
                     'isEOF': 1,
                     'nReturned': 5120,
                     'needTime': 435,
                     'needYield': 0,
                     'restoreState': 5,
                     'saveState': 5,
                     'stage': 'COLLSCAN',
                     'works': 5556},
 'executionSuccess': True,
 'executionTimeMillis': 46,
 'nReturned': 5120,
 'totalDocsExamined': 5555,
 'totalKeysExamined': 0}


In [51]:
db.transactions.create_index([('transactions.transactions.date', 1)])

'transactions.transactions.date_1'

In [52]:
find = db.transactions.find({'transactions.transactions.date': {'$gt': datetime(1997, 1, 1, 0, 0),'$lt': datetime(2001, 12, 31, 0, 0)}})
stats = find.explain()
pprint(stats['executionStats'])

{'allPlansExecution': [{'executionStages': {'advanced': 101,
                                            'alreadyHasObj': 0,
                                            'docsExamined': 101,
                                            'executionTimeMillisEstimate': 0,
                                            'filter': {'transactions.transactions.date': {'$lt': datetime.datetime(2001, 12, 31, 0, 0)}},
                                            'inputStage': {'advanced': 101,
                                                           'direction': 'forward',
                                                           'dupsDropped': 1,
                                                           'dupsTested': 102,
                                                           'executionTimeMillisEstimate': 0,
                                                           'indexBounds': {'transactions.transactions.date': ['(new '
                                                                     

In [53]:
db.transactions.drop_index('transactions.transactions.date_1')

In [54]:
#deciding the given date and periods of time:
id_property = '10006546'
start_date = datetime(1997, 1, 1, 0, 0)
end_date = datetime(2001, 12, 31, 0, 0)

#Creating the aggregation pipeline:
query_1 = {'$match': {'_id': id_property}}

query_2 = {'$unwind':'$transactions.transactions'}

query_3 = {'$match': {'transactions.transactions.date': {'$gt': start_date,'$lt': end_date}}}

query_4 = {'$group': {
            '_id':'$_id',
            'avg_price':{'$avg': {'$toDecimal':'$transactions.transactions.price'}}}
          }

query_5 = {'$project': {
            '_id':0,
            'avg_price':1,
            'property_id':'$_id'}
          }


pipeline = [query_1, query_2, query_3, query_4, query_5]

result = db.transactions.aggregate(pipeline)
pprint(list(result))

[{'avg_price': Decimal128('15.47911774898196268779315687424968'),
  'property_id': '10006546'}]


### Question 12

We wish to have a summary webpage that displays information about our top destinations. This webpage should display for each of the top 10 cities some basic information about our operations in the area (number of properties by type for example, average price by type) but you can choose the metrics. For each of the top 10 cities it should also provide some basic information about the top 3 properties in each city (price, number of review, whatever you think useful) to show an example of the properties available in the area. We would like to keep this webpage up to date as information changes.

Analysing the code below stage by stage:

* 1st `match`: We previously knew there was a city '', so in order to not count it, we had to match all documents that weren't equal to ''.

* 2nd `group`:  it's ID-ing by city and type of property, calculating the average price, total of properties and reviews.

* 3rd `group`: it's grouping by city, placing property_types in an array with the total of properties the average price. And grouping with the total number of reviews and the total properties by city.

* 4th `sort`: sorting by the descending order of total reviews by city, the previous output.

* 5th `limit`: limiting the output by ten, in order of getting the top ten cities. In here, we decided to use total reviews as the metric of defining what would make the top ten - popularity.

* 6th `lookup`: to get the top three properties with the most reviews, this stage is joining the top ten cities with the listings collection. The double $ is used when refering to a variable defined in previous stages of the aggregation pipeline.

* 7th `project`: this stage simply projects the output as we define it.

In [55]:
query_1 = {'$match': {'address.market': {'$ne': ''}}} 

query_2 = {'$group': {
            '_id': {'city': '$address.market', 'type': '$property_type'}, 
            'total_properties': {'$sum': 1},
            'total_reviews': {'$sum': {'$size': '$reviews'}}, 
            'average_price': {'$avg': '$price'}}
          }
    
query_3 = {'$group': {
            '_id': '$_id.city', 
            'property_types': {
                '$push': {
                    'type': '$_id.type', 
                    'count': '$total_properties',
                    'average_price':{'$round': [{'$avg': '$average_price'}, 2]}}},
            'total_reviews': {'$sum': '$total_reviews'},
            'total_properties': {'$sum': '$total_properties'}}
    }
    
query_4 = {'$sort': {'total_reviews': -1}}
    
query_5 = {'$limit': 10}
    
query_6 = {'$lookup': {
            'from': 'listings', 
            'let': {'city': '$_id'}, 
            'pipeline': [{
                '$match': {'$expr': {'$eq': ['$address.market', '$$city']}} 
            },{
                '$sort': {'number_of_reviews': -1}
            },{
                '$limit': 3
            },{
                '$project': {'_id': 1, 'name': 1, 'price': 1, 'number_of_reviews': 1}
            }],
            'as': 'top3_properties'}
    }

query_7 = {'$project': {
            '_id': 1,
            'total_reviews':1,
            'total_properties': 1,
            'property_types': 1, 
            'top3_properties': 1}
    }

query_8 = {'$merge': {
            'into': 'top10_cities',
            'whenMatched': 'replace'}
    }


pipeline = [query_1, query_2, query_3, query_4, query_5, query_6, query_7, query_8]

result = list(db.listings.aggregate(pipeline))

pprint(db.top10_cities.find_one())

{'_id': 'New York',
 'property_types': [{'average_price': Decimal128('135.57'),
                     'count': 489,
                     'type': 'Apartment'},
                    {'average_price': Decimal128('219.00'),
                     'count': 24,
                     'type': 'Townhouse'},
                    {'average_price': Decimal128('112.50'),
                     'count': 2,
                     'type': 'Guest suite'},
                    {'average_price': Decimal128('202.16'),
                     'count': 19,
                     'type': 'Condominium'},
                    {'average_price': Decimal128('139.42'),
                     'count': 19,
                     'type': 'Loft'},
                    {'average_price': Decimal128('100.00'),
                     'count': 1,
                     'type': 'Aparthotel'},
                    {'average_price': Decimal128('157.00'),
                     'count': 1,
                     'type': 'Hostel'},
                    {'aver

### Question 13
Add a new property, this property should have a new host, and should be located in one of the top 10 cities. The host selects the top 10 most common amenities to be listed for the property.

In [56]:
#Recalling the top10 cities:

query_1 = {'$match': {'address.market': {'$ne': ''}}}

query_2 = {"$group":{
                "_id":"$address.market",
                "number_of_properties":{
                    "$sum": 1}}
          }

query_3 = {"$project":{
                "_id":0,
                "city":"$_id",
                "number_of_properties":1}
          }

query_4 = {"$sort":{
                "number_of_properties":-1,
                "city":1}
          }

query_5 = {"$limit":10}


pipeline = [query_1, query_2, query_3, query_4, query_5]

result = list(db.listings.aggregate(pipeline))
pprint(result)

[{'city': 'Istanbul', 'number_of_properties': 660},
 {'city': 'Montreal', 'number_of_properties': 648},
 {'city': 'Barcelona', 'number_of_properties': 632},
 {'city': 'Hong Kong', 'number_of_properties': 619},
 {'city': 'Sydney', 'number_of_properties': 609},
 {'city': 'New York', 'number_of_properties': 607},
 {'city': 'Rio De Janeiro', 'number_of_properties': 603},
 {'city': 'Porto', 'number_of_properties': 554},
 {'city': 'Oahu', 'number_of_properties': 253},
 {'city': 'Maui', 'number_of_properties': 153}]


In [57]:
#Recalling the top 10 amenities:
pprint(list(db.amenities.find().limit(10)))

[{'_id': ObjectId('662598a65718344933ddc4de'),
  'amenityName': 'Wifi',
  'count': 5303},
 {'_id': ObjectId('662598a65718344933ddc4df'),
  'amenityName': 'Essentials',
  'count': 5048},
 {'_id': ObjectId('662598a65718344933ddc4e0'),
  'amenityName': 'Kitchen',
  'count': 4951},
 {'_id': ObjectId('662598a65718344933ddc4e1'),
  'amenityName': 'TV',
  'count': 4295},
 {'_id': ObjectId('662598a65718344933ddc4e2'),
  'amenityName': 'Hangers',
  'count': 4226},
 {'_id': ObjectId('662598a65718344933ddc4e3'),
  'amenityName': 'Hair dryer',
  'count': 3900},
 {'_id': ObjectId('662598a65718344933ddc4e4'),
  'amenityName': 'Washer',
  'count': 3877},
 {'_id': ObjectId('662598a65718344933ddc4e5'),
  'amenityName': 'Shampoo',
  'count': 3709},
 {'_id': ObjectId('662598a65718344933ddc4e6'),
  'amenityName': 'Iron',
  'count': 3692},
 {'_id': ObjectId('662598a65718344933ddc4e7'),
  'amenityName': 'Laptop friendly workspace',
  'count': 3442}]


Creating the document of a new property of a new host:

In [58]:
from bson import Decimal128

new_property = db.listings.insert_one({
    '_id': '00006540',
    'address': {
        'country': 'Portugal',
        'country_code': 'PT',
        'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, Vitória',
        'market': 'Porto',
        'street': 'Porto, Porto, Portugal'
    },
    'amenities': [
        'Wifi',
        'Essentials',
        'Kitchen',
        'TV',
        'Hangers',
        'Hair dryer',
        'Washer',
        'Shampoo',
        'Iron',
        'Laptop friendly workspace'
    ],
    'availability': {
        'availability_30': 20,
        'availability_365': 211,
        'availability_60': 51,
        'availability_90': 70
    },
     'host_information': {
            'id': '00000391',
            'is_superhost': False,
            'listings_count': 1,
            'name': 'João&Julia'
        },
    'maximum_nights': 20,
    'minimum_nights': 3,
    'name': 'Residência Dragão',
    'price': Decimal128('60.00'),
    'property_type': 'House'
})

In [59]:
#confirming it was indeed created:
document = db.listings.find({'name': 'Residência Dragão'})

for x in document:
    pprint(x)

{'_id': '00006540',
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal'},
 'amenities': ['Wifi',
               'Essentials',
               'Kitchen',
               'TV',
               'Hangers',
               'Hair dryer',
               'Washer',
               'Shampoo',
               'Iron',
               'Laptop friendly workspace'],
 'availability': {'availability_30': 20,
                  'availability_365': 211,
                  'availability_60': 51,
                  'availability_90': 70},
 'host_information': {'id': '00000391',
                      'is_superhost': False,
                      'listings_count': 1,
                      'name': 'João&Julia'},
 'maximum_nights': 20,
 'minimum_nights': 3,
 'name': 'Residência Dragão',
 'price': Decima

### Question 14

Add a new review to this property, the review should be from one of our top 20 reviewers.

In [60]:
#Getting the top 20 reviewer
query_1 = {"$project": {
                "_id": 1,
                "reviewer_name": 1,
                "number_of_reviews": 1}
          }
    
query_2 = {"$sort": {"number_of_reviews": -1}}
    
query_3 = {"$limit": 20}


pipeline = [query_1, query_2, query_3]

result = list(db.reviewers.aggregate(pipeline))
pprint(result)

[{'_id': '20775242', 'number_of_reviews': 24, 'reviewer_name': 'Filipe'},
 {'_id': '67084875', 'number_of_reviews': 13, 'reviewer_name': 'Nick'},
 {'_id': '2961855', 'number_of_reviews': 10, 'reviewer_name': 'Uge'},
 {'_id': '20991911', 'number_of_reviews': 9, 'reviewer_name': 'Lisa'},
 {'_id': '162027327', 'number_of_reviews': 9, 'reviewer_name': 'Thien'},
 {'_id': '69140895', 'number_of_reviews': 8, 'reviewer_name': 'Lisa'},
 {'_id': '12679057', 'number_of_reviews': 8, 'reviewer_name': 'Jodi'},
 {'_id': '55241576', 'number_of_reviews': 8, 'reviewer_name': 'Courtney'},
 {'_id': '1705870', 'number_of_reviews': 8, 'reviewer_name': 'David'},
 {'_id': '60816198', 'number_of_reviews': 8, 'reviewer_name': 'Todd'},
 {'_id': '78093968', 'number_of_reviews': 7, 'reviewer_name': 'David'},
 {'_id': '47303133', 'number_of_reviews': 7, 'reviewer_name': 'Lance'},
 {'_id': '24667379', 'number_of_reviews': 6, 'reviewer_name': 'Karen'},
 {'_id': '25715809', 'number_of_reviews': 6, 'reviewer_name': 'Me

Creating the new review made to the property created in the previous question:

In [61]:
date_now = datetime.now()

new_review = {
    'comments': 'We have loved to stay here, and the journey it took us. Best regards, IMS bit, SALLY.',
    'date': date_now,  
    '_id': '00006540',
    'reviewer_id': '20775242',
    'reviewer_name': 'Filipe', 
    'review_scores': {'checkin': 8,
                   'cleanliness': 9,
                   'communication': 4,
                   'location': 8,
                   'rating': 97,
                   'value': 7}
}

new_reviewer = {
    '_id': '20775242',
    'reviewer_name': 'Filipe',
    'number_of_reviews': 1,
    'reviews': [{
        'comment': 'We have loved to stay here, and the journey it took us. Best regards, IMS bit, SALLY.',
        'date': date_now,
        'listing_id': '00006540' 
    }]
}

db.listings.update_one(
    {'_id': '00006540'},  
    {'$push': {'reviews': new_review}})

db.reviews.insert_one(new_review)

#updating reviewers:
new_comment = {
    'comment': 'We have loved to stay here, and the journey it took us. Best regards, IMS bit, SALLY.',
    'date': datetime.now(),
    'listing_id': '00006540'
}


db.reviewers.update_one(
    {'_id': '20775242'},
    {
        '$set': {'number_of_reviews': 25},
        '$push': {'reviews': new_comment} 
    }
)

UpdateResult({'n': 1, 'nModified': 1, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)

In [62]:
#confirming it was created:
pipeline = [
    {
        '$match': {
            '_id': '00006540'
        }
    }
]

result = db.reviews.aggregate(pipeline)
pprint(list(result))

[{'_id': '00006540',
  'comments': 'We have loved to stay here, and the journey it took us. Best '
              'regards, IMS bit, SALLY.',
  'date': datetime.datetime(2024, 4, 21, 23, 52, 25, 856000),
  'review_scores': {'checkin': 8,
                    'cleanliness': 9,
                    'communication': 4,
                    'location': 8,
                    'rating': 97,
                    'value': 7},
  'reviewer_id': '20775242',
  'reviewer_name': 'Filipe'}]


### Question 15

Add a new review metric called x_factor to the new document and give a score of 10. Show that the average across all metrics is correctly calculated for this listing, with the query you previously developed.

In [63]:
new_metric={'$set': {'review_scores.x_factor': 10}}

review_id={'_id': '00006540'}

db.reviews.update_one(review_id, new_metric)

UpdateResult({'n': 1, 'nModified': 1, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)

In [64]:
pprint(list(db.reviews.find({'_id':'00006540'})))

[{'_id': '00006540',
  'comments': 'We have loved to stay here, and the journey it took us. Best '
              'regards, IMS bit, SALLY.',
  'date': datetime.datetime(2024, 4, 21, 23, 52, 25, 856000),
  'review_scores': {'checkin': 8,
                    'cleanliness': 9,
                    'communication': 4,
                    'location': 8,
                    'rating': 97,
                    'value': 7,
                    'x_factor': 10},
  'reviewer_id': '20775242',
  'reviewer_name': 'Filipe'}]


In [65]:
property_id = '00006540'

pipeline_10 = [
    {'$match': {'_id': property_id}},
    {'$project': {
        '_id': 1,
        'score_array': {'$objectToArray': '$review_scores'}
    }},
    {'$project': {
        '_id': 1,
        'average_score': {'$avg': '$score_array.v'}
    }}
]

result = list(db.reviews.aggregate(pipeline_10))
pprint(result)

[{'_id': '00006540', 'average_score': 20.428571428571427}]
